src/main/kotlin/com/hj/leetcode/kotlin/problem1489/Solution.kt

```diff
@@ -0,0 +1,140 @@
+package com.hj.leetcode.kotlin.problem1489
+
+/**
+ * LeetCode page: [1489. Find Critical and Pseudo-Critical Edges in Minimum Spanning Tree](https://leetcode.com/problems/find-critical-and-pseudo-critical-edges-in-minimum-spanning-tree/);
+ */
+class Solution {
+    /* Complexity:
+     * Time O(E^2 * Log n) and Space O(E) where E is the size of edges;
+     */
+    fun findCriticalAndPseudoCriticalEdges(n: Int, edges: Array<IntArray>): List<List<Int>> {
+        val sortedEdges = sortedEdges(edges) // Sorted by monotonically increasing weight
+        val mstWeight = mstWeight(n, sortedEdges)
+
+        val criticalEdges = criticalEdges(n, sortedEdges, mstWeight)
+        val pseudoCriticalEdges = pseudoCriticalEdges(n, sortedEdges, criticalEdges, mstWeight)
+
+        return listOf(criticalEdges.map { it.index }, pseudoCriticalEdges.map { it.index })
+    }
+
+    private data class Edge(val u: Int, val v: Int, val weight: Int, val index: Int)
+
+    private fun sortedEdges(edges: Array<IntArray>): List<Edge> {
+        val result = edges.mapIndexedTo(mutableListOf()) { index, array ->
+            Edge(array[0], array[1], array[2], index)
+        }
+        result.sortBy { edge -> edge.weight }
+        return result
+    }
+
+    private fun mstWeight(n: Int, sortedEdges: List<Edge>): Int {
```

Comment:
Most of the logic for finding the MST weight is shared among this function, `criticalEdges(...)`, and `pseudoCriticalEdges(...)`. By adding functionality to handle edges that must be included or excluded, we can reduce duplicated code.

Reply:
Resolved by commints: https://github.com/hj-core/hj-leetcode-kotlin/pull/130/commits/d528c41b2a57454628b25254251f904d80ff81ee, https://github.com/hj-core/hj-leetcode-kotlin/pull/130/commits/18d7783c45411c9ed448a87dba0319f21c288221, https://github.com/hj-core/hj-leetcode-kotlin/pull/130/commits/195826bf3aa290c4c0f43600493a49bfdd524bc8, https://github.com/hj-core/hj-leetcode-kotlin/pull/130/commits/4234ce542db5e3226f846e41f28aabafb8698f00